patches/api/0018-introduce-a-concept-of-disguise-that-is-paper-prone.patch

```diff
@@ -0,0 +1,144 @@
++import org.jetbrains.annotations.NotNull;
++
++public record EntityTypeDisguise(@NotNull EntityType entityType) implements DisguiseData {
++    public EntityTypeDisguise {
++        Objects.requireNonNull(entityType, "type cannot be null");
++    }
++}
+diff --git a/src/main/java/com/github/weepingmc/disguise/PlayerDisguise.java b/src/main/java/com/github/weepingmc/disguise/PlayerDisguise.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..b8525cc42825c18d49e07f89f219a68ad7ea39c2
+--- /dev/null
++++ b/src/main/java/com/github/weepingmc/disguise/PlayerDisguise.java
+@@ -0,0 +1,33 @@
++package com.github.weepingmc.disguise;
++
++import com.destroystokyo.paper.profile.PlayerProfile;
++import java.util.Objects;
++import org.jetbrains.annotations.NotNull;
++
++public record PlayerDisguise(@NotNull PlayerProfile playerProfile, boolean listed) implements DisguiseData {
++
++    public  PlayerDisguise {
++        Objects.requireNonNull(playerProfile, "profile cannot be null");
++    }
++    public static @NotNull Builder builder(@NotNull PlayerProfile playerProfile) {
++        return new Builder(playerProfile);
++    }
++
++    public static class Builder {
++        private final PlayerProfile playerProfile;
++        private boolean listed;
++
++        public Builder(@NotNull PlayerProfile playerProfile) {
++            this.playerProfile = playerProfile;
++        }
++
++        public @NotNull Builder listed(boolean listed) {
++            this.listed = listed;
++            return this;
++        }
++
++        public @NotNull PlayerDisguise build() {
++            return new PlayerDisguise(playerProfile, listed);
++        }
++    }
++}
+diff --git a/src/main/java/com/github/weepingmc/event/packet/PlayerTrackEntityEvent.java b/src/main/java/com/github/weepingmc/event/packet/PlayerTrackEntityEvent.java
+index 7f64273cbf447ad9dcd73512c1936bd5d173e032..392c1e03e6a93170eb4666fa7ac422c22959b988 100644
+--- a/src/main/java/com/github/weepingmc/event/packet/PlayerTrackEntityEvent.java
++++ b/src/main/java/com/github/weepingmc/event/packet/PlayerTrackEntityEvent.java
+@@ -1,5 +1,7 @@
+ package com.github.weepingmc.event.packet;
+ 
++import com.github.weepingmc.disguise.DisguiseData;
++import java.util.Objects;
+ import org.bukkit.entity.Entity;
+ import org.bukkit.entity.Player;
+ import org.bukkit.event.Cancellable;
+@@ -7,6 +9,7 @@ import org.bukkit.event.Event;
+ import org.bukkit.event.HandlerList;
+ import org.bukkit.event.player.PlayerEvent;
+ import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
+ 
+ /**
+  * Is called before a {@link Player} tracks an {@link Entity}.
+@@ -25,6 +28,7 @@ public class PlayerTrackEntityEvent extends PlayerEvent implements Cancellable {
+     private static final HandlerList handlers = new HandlerList();
+     private final Entity entity;
+     private boolean cancelled;
++    private DisguiseData disguiseData = DisguiseData.reset();
```

Comment:
Decision if we should it on per event basis or attached onto the entity.

If we add it to the entity we won't need an addition event for clientbound entity meta -> to filter out the invalid meta for that type.